### ISSUE TITLE
non-file multi tasks crash with --verbose

### ISSUE THREAD

**Issue by marneborn** [gruntjs/grunt#1341]
When I register a multi task that isn't file based grunt crashes with --verbose.
--force doesn't help.

Using the example multitask from the doc:

```
grunt --verbose --stack --force log:baz
Initializing
Command-line options: --verbose, --stack, --force

Reading "Gruntfile.js" Gruntfile...OK

Registering Gruntfile tasks.
Initializing config...OK
Loading "Gruntfile.js" tasks...OK
+ log

Running tasks: log:baz

Running "log:baz" (log) task
Verifying property log.baz exists in config...OK
Warning: Object false has no method 'indexOf' Used --force, continuing.

Done, but with warnings.
```

**Comment by silkentrance** [comment]
The problem lies in registerMultiTask and normalizeMultiTaskFiles which tries to interpret the specified target, in this case baz or foo, which is an array of numbers, as string patterns.

Looking at https://github.com/gruntjs/grunt/blob/master/lib/grunt/task.js#L121 ff. it seems as if grunt is a bit biased in that it assumes that everything from that target is also a valid path or pattern.

Since the target could be an array of strings or a combination of strings and other non string objects and perhaps even deeply nested so, I don`t know how to fix this without possibly breaking existing gruntfiles.

Any ideas @cowboy @tkellen?

**Comment by matmar10** [comment]
Ping

**Comment by shama** [comment]
I'm not understanding the issue. Would you be able to create an example I could run that demonstrates the problem?

**Comment by matmar10** [comment]
The issue it seems is using --verbose with a non-object target property:
```
  grunt.initConfig({
    example: {
      target: true, // this fails
      otherTarget: {
        // this is OK
      }
    }
  });
// grunt example:target ~> OK
// grunt example:otherTarget ~> OK
// grunt example:target —verbose ~> FAILS
// grunt example:otherTarget —verbose ~> OK
```
Full working (breaking) example here: [https://github.com/matmar10/grunt-verbose-fails-example](https://github.com/matmar10/grunt-verbose-fails-example)

**Comment by shama** [comment]
Thanks @matmar10!

I'm curious why non-objects are being supplied to multi tasks? This seems like a user error to me. We certainly can improve the error but I just want to be sure we don't add code that supports a behavior we don't intend on supporting.

**Comment by matmar10** [comment]
@shama The use case for us is "for this environment, run or do not run this part of the task". Or for example, just need a URL per-environment (e.g. dev -> dev.app.com, prod -> prod.app.com).

**Comment by shama** [comment]
@matmar10 I appreciate you sharing!

`grunt.registerMultiTask()` is intended to be used with `this.files` and we have considered removing support for `this.data` in those tasks. So it's good to hear about use cases.

A better alternative for tasks which don't supply `src/dest` blocks of files is:

```js
grunt.registerTask('task', function (target) {
  var data = grunt.config([this.name, target]);
});
```

Similar to what is done in the [watch task](https://github.com/gruntjs/grunt-contrib-watch/blob/master/tasks/watch.js#L60). This way we don't accidentally break custom task behaviors when updating how multi-tasks normalize files.

**Comment by marneborn** [comment]
@shama thanks for the clarification. With the munging that happens during registerMutliTask, using grunt.config is definitely safer (eg `{target : [['a', 'b'],['c', 'd']]}` => `this.data == ['a','b','c','d']`).

However, the multi tasks example on the [web site](http://gruntjs.com/creating-tasks#multi-tasks) doesn't reflect that intended use. From that example, log:foo and log:baz both have this issue.
```
grunt.initConfig({
  log: {
    foo: [1, 2, 3],
    bar: 'hello world',
    baz: false
  }
});

grunt.registerMultiTask('log', 'Log stuff.', function() {
  grunt.log.writeln(this.target + ': ' + this.data);
});
```

To expand a little on the problem, it's not simply non-objects that cause this error. Rather, it's all non-strings in the target.src created during registration.  So all of these are a problem:

```javascript
  grunt.initConfig({
    test: {
      case1: 1,
      case2: [false],
      case3: [{key:"value"}]
    }
  });
```

When I found this I was trying to use the multiTask to pick arguments to apply to a function. Though I don't do it this way anymore.

Something like this:
```javascript
module.exports = function(grunt) {
  grunt.initConfig({
    multi: {
      task1: ['a', 1],
      task2: ['b', 2]
    }
  });
  grunt.registerMultiTask('multi', function() {
    somefunction.apply(null, this.data);
  });
};
```

**Comment by ljqx** [comment]
@shama when I'm fixing a bug in `grunt-webpack`, one test fails because of this issue. I think this could be a real life example.

With same configuration the test will pass if no `--verbose` option provided.

**Comment by matmar10** [comment]
I agree with @ljqx that at a minimum, the behavior should be consistent between `--verbose` and non `--verbose` since the difference in behavior breaks the APIs implicit contract of running with verbose gives more logging output.
